core/headerchain.go

```diff
@@ -460,13 +477,18 @@ func (hc *HeaderChain) CurrentHeader() *types.Header {
 
 // SetCurrentHeader sets the current head header of the canonical chain.
 func (hc *HeaderChain) SetCurrentHeader(head *types.Header) {
-	rawdb.WriteHeadHeaderHash(hc.chainDb, head.Hash())
-
 	hc.currentHeader.Store(head)
 	hc.currentHeaderHash = head.Hash()
 	headHeaderGauge.Update(head.Number.Int64())
 }
 
+// WriteHeadHeader sets the current head header of the canonical chain
+// and write the marker into database.
+func (hc *HeaderChain) WriteHeadHeader(db ethdb.KeyValueWriter, head *types.Header) {
+	rawdb.WriteHeadHeaderHash(db, head.Hash())
+	hc.SetCurrentHeader(head)
```

Comment:
This should probably not be done here (while still only writing to batch), but done by caller after the batch is committed